### ISSUE TITLE
TaggedIOException: unsupported feature (data descriptor + stored flags) could cause internal ZIP structure to be lost

### ISSUE THREAD

**Issue by tc-wleite** [sepinf-inc/IPED#655]
Another person here is working in a case with a single important file, which is a **ZIP** containing many files.
Analysing the processed content, the user noticed that the ZIP internal structure was partially lost.
Some files were processed correctly, but many others were placed as they were all in the ZIP root level, although they weren't.
No files seem to be missing, just part of them were mixed together in the root folder.

I was able to reproduce the problem (both in 3.18.x and master), and checked that the file opens without any error/warning using 7-Zip or the Windows integrated ZIP expander.  

I noticed the following message in the processing log:
`Error exporting [file.zip>>folder/file.ext]  org.apache.tika.io.TaggedIOException: unsupported feature data descriptor used in entry [folder/file.ext]`
(actual files and folder names were suppressed).

Digging into this issue, it seems to be happening inside a library (package `org.apache.commons.compress.archivers`) used by `PackageParser`.
I found a possible workaround (adding a few lines in `PackageParser`) that solved the issue for this file.
I still need to: 
- Test with other ZIP (and other archive types handled by this parser) and check if there are no side effects.
- Find other files that trigger this error and see if the fix handles them too. 

The file that caused this issue contains illegal/sensitive content, so I can't share it here.
Taking a quick look in recent processing logs from other cases, I found a few identical warning messages, so it should  be possible to find more of these ZIPs, although it seems to be a rare issue.

**Comment by lfcnassif** [comment]
Thanks @tc-wleite for reporting this and for digging into it! Let me know if you need any help.

**Comment by lfcnassif** [comment]
Hey @tc-wleite if you would like I can take care of this if you send me the triggering file(s) and the proposed fix, thanks.

**Comment by tc-wleite** [comment]
> Hey @tc-wleite if you would like I can take care of this if you send me the triggering file(s) and the proposed fix, thanks.

I worked a bit on this last week...
The fix I tried worked on the file that I had, but when I tested with a larger set, there were other files negatively affected by the change, so it is not a possible solution.
Maybe you can figure it out, as it involves parts of the code that you are more familiar with.
I will try to find example files that can be shared here, otherwise I will sent you privately the initial one.

**Comment by tc-wleite** [comment]
In my tests, out of ~30,000 ZIP files from many cases, about 100 of them triggered the reported issue.
All these ~100 files are expanded by Windows and 7-ZIP without any problem.

I found that this small file should cause the exception:
[qr.zip](https://github.com/sepinf-inc/IPED/files/6945529/qr.zip)
Not sure if it is the best example, but the others I found have sensitive content.

The possible fix that I found was to enable "allowStoredEntriesWithDataDescriptor" (which is disabled in the used constructor) when creating the ZipArchiveInputStream.
In PackageParser:
```
ais = factory.createArchiveInputStream(new CloseShieldInputStream(stream));
// Added lines:
if (ais instanceof ZipArchiveInputStream)
    ais = new ZipArchiveInputStream(new CloseShieldInputStream(stream), "UTF8", true, true);
```

It fixed the problem for the original file that I identified, but has no effect or created new issues with other files, so it is not an option (at least in this simple way).

I think the problem came from known limitations of handling ZIP files as streams (ZipArchiveInputStream), as described here:
http://commons.apache.org/proper/commons-compress/zip.html#ZipArchiveInputStream_vs_ZipFile

**Comment by lfcnassif** [comment]
Thanks for the sample.

> I think the problem came from known limitations of handling ZIP files as streams (ZipArchiveInputStream), as described here:

Yes, I was aware about some of those limitations, but parsing ZIPs as streams allows to extract entries from corrupted/carved ZIP files where the central directory was overwritten (a lot of corrupted ZIPs aren't expanded by Windows Explorer and rarely by 7zip, but IPED does). Not sure if switching to ZipFile is the better approach...

But, if an exception is always thrown (right?) I think it could be catched and we could maybe use the alternativeParse(...) method, I'll do some testing with the sample.

**Comment by tc-wleite** [comment]
I also think switching to ZipFile is not a solution, but maybe there is a way to use alternative approaches when an exception is thrown as you pointed out.
Right, an exception is thrown. A TaggedIOException for the initial file that motivated this issue (and the other ~100 files that I mentioned). 

I saw other exceptions in the log when processing other ZIP files I collected here, some of them may be related to the limitations of ZipArchiveInputStream, but probably most were just corrupted files.
A good start would be handling TaggedIOException (detecting and handle the file in a different way). 
Later other exception could be identified and expand this solution.

**Comment by lfcnassif** [comment]
@tc-wleite I pushed a non ideal first attempt to fix this (it is duplicating the directory entry, but maybe better than other errors...)

If you could test with other triggering files, let me know the result, or please send them privately to me.

Tika recently fixed a very similar issue, and they used some code to count the number of items extracted and skip them later in a second parse using the flag you tried before, maybe we can use some similar approach, what do you think?
https://github.com/apache/tika/pull/364

**Comment by lfcnassif** [comment]
Just tried the new Tika approach in a new branch (commit 97954ee), but now the sample throws:

org.apache.tika.io.TaggedIOException: compressed and uncompressed size don't match while reading a stored entry using data descriptor. Either the archive is broken or it can not be read using ZipArchiveInputStream and you must use ZipFile. A common cause for this is a ZIP archive containing a ZIP archive. See http://commons.apache.org/proper/commons-compress/zip.html#ZipArchiveInputStream_vs_ZipFile

So I just simplified the initial approach with 052a435, but it still produces duplicated entries.

**Comment by lfcnassif** [comment]
Another idea is iterating over N first zip entries and checking if STORED + DATA_DESCRIPTOR flags are used, without extracting the files. If not, reset/get a new stream and parse normally. If yes, use the SevenZip alternative parsing. Not sure if this would cause performance regression issues if the first entries are big files.

**Comment by lfcnassif** [comment]
And I think it is possible to have mixed entries in the same zip file, with and without those flags, but maybe that is not common out there.

**Comment by lfcnassif** [comment]
I just had an idea to filter out duplicated entries, maybe we can inject an EmbeddedDocumentExtractor proxy in ParseContext from PackageParser to filter out already extracted files when redirecting to SevenZipParser in this line https://github.com/sepinf-inc/IPED/blob/master/iped-parsers/iped-parsers-impl/src/main/java/dpf/sp/gpinf/indexer/parsers/SevenZipParser.java#L278

**Comment by lfcnassif** [comment]
Just pushed the last approach. If you could send me privately the other triggering files, that would be great.

**Comment by tc-wleite** [comment]
> Just pushed the last approach. If you could send me privately the other triggering files for testing, that would be great.

Just started uploading a set of ~20 ZIP files (in total ~4.5 GB) that should cause the TaggedIOException (to /Temp/Wladimir).
It will take another 40 minutes or so to finish the upload.

**Comment by lfcnassif** [comment]
Just downloaded the files, you can delete them.

**Comment by lfcnassif** [comment]
I have some news:
- Subitem path issue actually is a bug in SevenZipJBinding used by SevenZipParser. I tested the old version used before #282 update and the issue was already there. Fortunatelly we don't use SevenZipParser for most ZIP files, just when this exception is thrown.
- So I changed the alternative parsing method, called after UnsupportedZipFeatureException is thrown because of data_descriptor + stored flags,  to use commons compress ZipFile instead of SevenZip, and the internal path issue was gone with the samples @tc-wleite sent to me.
- With that, now duplicated entries are being filtered fine, because it is based on subitem paths

As the new code is just used when the exception is thrown, regular zip files, at first, wouldn't be affected. But this needed a minor refactor because ZipFile and ZipArchiveInputStream APIs are different. I will do a final review before merging.

@tc-wleite if you could test with your other samples, that would be great!

**Comment by tc-wleite** [comment]
Excellent @lfcnassif!
 
> @tc-wleite if you could test with your other samples, that would be great!

Sure! I can run with a larger test set.
Should I get the modified version from [655_fix branch](https://github.com/sepinf-inc/IPED/tree/%23655_fix)?

**Comment by lfcnassif** [comment]
> Should I get the modified version from [655_fix branch](https://github.com/sepinf-inc/IPED/tree/%23655_fix)?

Yes, when you have time, thanks.

**Comment by lfcnassif** [comment]
Just to complement, I just found a minor regression with DIRAC 8 RF.zip (22262300 bytes) sample. Previous code used to extract a zero sized pdf file, new code throws a corrupted zip error. Windows is not able to open that zip. SevenZip does, but results in an error when trying to extract the PDF. So we are just loosing the PDF file name from a corrupted zip. Other samples I tested seems good to me.

**Comment by lfcnassif** [comment]
Just checked a small test with 7k zips (25GB) from a single case, no regressions detected.

**Comment by tc-wleite** [comment]
A quick update here: 
The larger test is still running. As it is processing ~700 GB of ZIP files, it will take a while.
When it is done (later today), I will make a comparison with the same folder processed with the current version.
I will probably be able to update results on the beginning of the next week.

**Comment by lfcnassif** [comment]
Thanks @tc-wleite for all your time.

**Comment by tc-wleite** [comment]
@lfcnassif, compared the results of the tests (branch with your fix vs 3.18.7).
I checked the overall number of expanded itens and went into more detail for ~100 known problematic ZIPs I had identified before and a few other randomly selected  ZIPs. 
The fix solved the reported issue for all the files I have here that triggered it, and I didn't identify any side effect in other files.

**Comment by lfcnassif** [comment]
Great, thank you very much @tc-wleite! I know manual comparisons take a lot of time...
